docs/modules/ROOT/pages/references/style-guide.adoc

```diff
@@ -0,0 +1,273 @@
+  They draw attention to the wrong parts of the code and result in hard to review pull requests due to the white space changes.
++
+[source,jsonnet]
+----
+// Don't align vertically
+local plus     = "+";
+local minus    = "-";
+local multiply = "*";
+
+// Do the following
+local plus = "+";
+local minus = "-";
+local multiply = "*";
+----
+
+=== Blank Lines (Vertical Whitespace)
+
+* A single blank line appears:
+** Within method bodies, as needed to create logical groupings of statements.
+** Optionally before the first member or after the last member of a class or method.
+* Use one or two blank line(s) to separate class definitions.
+* Excessive number of blank lines is discouraged.
+
+=== Defining and Using Abstractions
+
+==== Defining Classes
+
+* Rather than defining a concrete JSON file, it's often useful to define a template which takes some set of parameters before being materialized into JSON.
+  We can liken named functions which take a set of parameters and result in a fixed scheme to ``classes'' in object-oriented languages, and so we will use that terminology.
+* When defining a class, use the following syntax:
++
+[source,jsonnet]
+----
+local newAnimal(name, age) = {
+  name: name,
+  age: age,
+};
+{
+  newAnimal:: newAnimal,
+}
+----
+
+* Returning a dictionary with a `newXxx` method (rather than just returning the constructor directly) allows exposing constants, static methods, or related class constructors from the same file.
+  In other words, it allows extending this class in the future without refactoring all downstream consumers.
+* When defining a class with both required and optional parameters, put required parameters first.
+  Optional parameters should have a default, or `null` if a sentinel value is needed.
++
+[source,jsonnet]
+----
+local newAnimal(name, age, isCat = true) = { ... }
+----
+
+* Wrap parameter declarations by putting one per line with 2 extra spaces of indentation, to differentiate from the method body.
+  Doing this is always acceptable, even if the definition would not wrap.
++
+[source,jsonnet]
+----
+local newAnimal(
+    name,
+    age,
+    isCat = true) = {
+  name: name,
+  …
+}
+----
+
+== Defining Methods
+
+* Method definitions follow the same syntactic style as class definitions.
+* Methods defined within a class should always be defined with `::`, as they fail to render with `:`.
+* Methods which return single values (rather than a dictionary) should use parentheses `()` to enclose their bodies if they're multi-line, identically to how braces would be used.
++
+[source,jsonnet]
+----
+{
+  multiply:: function(number1, number2): (
+    number1 * number 2
+  ),
+}
+----
+
+=== Using Classes
+
+* Import all dependencies at the top of the file and given them names related to the imported file itself.
+  This makes it easy to see what other files you depend on as the file grows.
++
+[source,jsonnet]
+----
+// CORRECT
+local animalTemplate = import "animal.jsonnet.TEMPLATE";
```

Comment:
what's up with the `.TEMPLATE` in the import here?

Shouldn't this be `import "animal.libjsonnet"` to make the style guide adhere to itself? We state later on that libraries should have extension `.libjsonnet`.